### PR TITLE
Feature/feat 3 UI and authentication adjustments

### PR DESCRIPTION
**Pull request by Peregrinus0304**
Please, take a look also on my CI - https://app.bitrise.io/app/15f132aede66f9d7

### CODE REVIEW

File: FantasyUniverse/Customization/Animations.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+//  Animations.swift
+//  FantasyUniverse
+//
+//  Created by TarasPeregrinus on 01.08.2022.
+//
+
+import SwiftUI
+
+extension View {
+    func animateWithSpring(
```
Comment (dzoanb) [inline FantasyUniverse/Customization/Animations.swift]: Maybe to indicate this happens on appear it's worth including this info in the name?

File: FantasyUniverse/Customization/ViewModifiers.swift [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    
+    func body(content: Content) -> some View {
+        content
+            .font(.appLargeFont)
+            .minimumScaleFactor(0.1)
+            .padding(.horizontal, 2)
+            .frame(width: width,
+                   height: height,
+                   alignment: .center)
+            .background(Color(Asset.Colors.aquamarine.color))
```
Comment (dzoanb) [inline FantasyUniverse/Customization/ViewModifiers.swift]: Wouldn't it be nicer to read and write if you add a background method that takes the `Asset.Colors` as an argument or even add an enum `AppColor` with constants of type `Color`, then you could call it like `.background(AppColor.aquamarine)

File: FantasyUniverse/Customization/ViewModifiers.swift [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+//  FantasyUniverse
+//
+//  Created by TarasPeregrinus on 29.07.2022.
+//
+
+import SwiftUI
+import Combine
+import UIKit
+
+struct DefaultLabelStyle: ViewModifier {
```
Comment (dzoanb) [inline FantasyUniverse/Customization/ViewModifiers.swift]: These Labels look awfully similar, wouldn't it be better to have one and parametrise it a bit more or having others use this one for example?

File: FantasyUniverse/Customization/ViewModifiers.swift [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+            .overlay(
+                RoundedRectangle(
+                    cornerRadius: UIValues.defaultCornerRadius)
+                    .stroke(Color(Asset.Colors.aquamarine.color),
+                            lineWidth: UIValues.defaultBorderWidth)
+            )
+    }
+}
+
+struct KeyboardAwareModifier: ViewModifier {
```
Comment (dzoanb) [inline FantasyUniverse/Customization/ViewModifiers.swift:78]: A good article, I know it 😄 I'd recommend adding a link in a comment to this class for later reference.

File: FantasyUniverse/Customization/ViewModifiers.swift [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            .frame(height: UIValues.defaultMinorElementHeight)
+            .padding(.horizontal, UIValues.defaultPadding)
+            .background(
+                LinearGradient(gradient:
+                                Gradient(colors:
+                                    [Color(Asset.Colors.aquamarine.color),
+                                     Color(Asset.Colors.appLint.color)]),
+                               startPoint: .top,
+                               endPoint: .bottom)
+                    .edgesIgnoringSafeArea(.all))
```
Comment (dzoanb) [inline FantasyUniverse/Customization/ViewModifiers.swift]: I'd rather expect this on the outside view, I don't think a background can go outside of safe area when the view itself does not 🤔

File: FantasyUniverse/Customization/ViewModifiers.swift [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        }
+    }
+    
+    func makeBody(configuration: Configuration) -> some View {
+        configuration.label
+            .font(.appMediumFont)
+            .minimumScaleFactor(0.1)
+            .padding(.horizontal, 2)
+            .frame(width: width,
+                   height: height)
```
Comment (dzoanb) [inline FantasyUniverse/Customization/ViewModifiers.swift:157]: In all these I see the width, but shouldn't it be enough to set it on the call site instead of having additional code to pass it through initialisation? In a nutshell, any parent view can set it on the  button that uses this style

File: FantasyUniverse/Models/Auth/User.swift [hunk 7]
```diff
@@ -7,18 +7,67 @@
 
 import Foundation
 import Combine
+import UIKit
 
-class User: ObservableObject {
+final class User: ObservableObject {
```
Comment (dzoanb) [inline FantasyUniverse/Models/Auth/User.swift:12]: I'm wondering if it could be beneficial to have User as a struct with costant properties and having something like `UserService` to actually manage it's lifecycle, handle observing and for example when you download the image you just return a new immutable User instance. Having all these properties mutable is dangerous for global mutable state and its issues.

File: FantasyUniverse/Models/Auth/User.swift [hunk 8]
```diff
@@ -7,18 +7,67 @@
...
+        self.lastName = lastName
+        
+        if let url = profileImageURL {
+            fetchedImagePublisher(from: url)
+                .receive(on: DispatchQueue.main)
+                .sink { res in
+                    switch res {
+                    case .failure(let err):
+                            debugPrint(err.localizedDescription)
+                    default: break
```
Comment (dzoanb) [inline FantasyUniverse/Models/Auth/User.swift:46]: Maybe worth to just handle the second case here?

File: FantasyUniverse/Scenes/Dashboard/DashboardView.swift [hunk 9]
```diff
@@ -9,37 +9,26 @@ import SwiftUI
...
-    @State var show = false
-    @State var correct = 0
-    @State var wrong = 0
-    @State var answered = 0
-    @State var selectedSet = ""
     let sets = QuestionSet.allSets
-    
+    @State var showSelectedQuiz = false
+    @State var showListItems = false
+    @State var animationDelay = 0.2
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Dashboard/DashboardView.swift]: This doesn't seem like something something that changes dynamically and then it doesn't have to be marked as State

File: FantasyUniverse/Scenes/Dashboard/DashboardView.swift [hunk 10]
```diff
@@ -9,37 +9,26 @@ import SwiftUI
...
+    @State var selectedSet: QuestionSet = .none
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Dashboard/DashboardView.swift]: Maybe a ViewModel makes sense?

File: FantasyUniverse/Scenes/Dashboard/DashboardView.swift [hunk 11]
```diff
@@ -59,38 +48,37 @@ struct DashboardView: View {
...
+                        selectedSet = sets[index]
+                        showSelectedQuiz.toggle()
                     })
                 }
             })
                 .frame(maxWidth: UIScreen.main.bounds.width - 30)
                 .padding()
         }
+        .onAppear {
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.2, execute: {
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Dashboard/DashboardView.swift]: Maybe some consts for this kind of const here? Maybe it should be equal to `animationDelay `?

File: FantasyUniverse/Scenes/Dashboard/DashboardViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+//  DashboardViewModel.swift
+//  FantasyUniverse
+//
+//  Created by TarasPeregrinus on 06.08.2022.
+//
+
+import Foundation
+import SwiftUI
+
+class DashboardViewModel: ObservableObject {}
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Dashboard/DashboardViewModel.swift]: Oh, I found it, now just needs populating 😄

File: FantasyUniverse/Scenes/ManualSignIn/ManualSignInView.swift [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            textFieldValue: $viewModel.loginFieldValue, valueModified: {
+                viewModel.validateCredentials()
+            })
+    }
+    
+    private var passwordField: some View {
+        CredentialsInputField(
+            labelText: "Password",
+            textFieldPlaceholder: "Type your password",
+            textFieldValue: $viewModel.passwordFieldValue, valueModified: {
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/ManualSignIn/ManualSignInView.swift:58]: If you keep the styling with all params in new lines then apply it everywhere

File: FantasyUniverse/Scenes/ManualSignIn/ManualSignInView.swift [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+                viewModel.validateCredentials()
+            })
+    }
+    
+    private var submitButton: some View {
+        Button("Sign in") {
+            if viewModel.fieldsValid {
+                viewModel.signIn()
+            }
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/ManualSignIn/ManualSignInView.swift:67]: These bits of logic belong to the view model, so it should be a part of the `signIn` method. Or maybe from the user perspective having the button visibly disabled and enabled dynamically could be nicer.

File: FantasyUniverse/Scenes/ManualSignIn/ManualSignInViewModel.swift [hunk 15]
```diff
@@ -44,11 +44,12 @@ class ManualSignInViewModel: ObservableObject {
             password: passwordFieldValue)
         firebaseAuthService
             .loginPublisher(with: credentials)
+            .receive(on: RunLoop.main)
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/ManualSignIn/ManualSignInViewModel.swift]: I'd prefer `DispatchQueue.main` to look less scary 😅

File: FantasyUniverse/Scenes/Quiz/QuizView.swift [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+struct QuizView: View {
+    
+    @Binding var set: QuestionSet
+    @StateObject var viewModel = QuizViewModel()
+    @Environment(\.presentationMode) var present
+    
+    var body: some View {
+        
+        AnimatedBackground(animationName: "day-background") {
+            if viewModel.questionsData.isEmpty {
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Quiz/QuizView.swift:20]: Maybe some states like `loading`, `empty`, `presenting(QuestionSet)` could be better for readability? It's semantically hard to follow that empty set means loading

File: FantasyUniverse/Scenes/Quiz/QuizView.swift [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+//  FantasyUniverse
+//
+//  Created by TarasPeregrinus on 25.12.2021.
+//
+
+import SwiftUI
+
+struct QuizView: View {
+    
+    @Binding var set: QuestionSet
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Quiz/QuizView.swift]: If you have it in the View model you shouldn't keep it here to keep a single source of truth rule

File: FantasyUniverse/Scenes/Quiz/QuizView.swift [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        let fraction = CGFloat(viewModel.answered) / CGFloat(viewModel.questionsData.count)
+        let width = UIScreen.main.bounds.width - 30
+        return fraction * width
+    }
+    
+    private func cleanUp() {
+        viewModel.correct = 0
+        viewModel.wrong = 0
+        viewModel.answered = 0
+    }
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Quiz/QuizView.swift]: Belongs to view model most definitely

File: FantasyUniverse/Scenes/Quiz/QuizView.swift [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+                }
+            }
+            
+        }
+        .navigationBarHidden(true)
+        .onAppear {
+            viewModel.getQuestions(set: set)
+        }
+        .onDisappear {
+            cleanUp()
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/Quiz/QuizView.swift]: You shouldn't need to clean it up, the instance should be deallocated for you on disappear

**Review by dzoanb — COMMENTED** [review]
Half done 😄

File: FantasyUniverse/Customization/ViewModifiers.swift (hunk 2 above)
Comment (Peregrinus0304) [reply FantasyUniverse/Customization/ViewModifiers.swift]: It would be nice, although i have not found a way to do it. Asset.Colors is an enum that contains static lets, not cases, so You can not add Asset.Colors as a type for parameter in custom init. You will not be able to use dot notation to call .colorOne from inside a custom a background method, because colors are lets, not cases. We have faced the same issue on our previous project, while implementing Tuist (because it uses the same mechanism and almost the same generated code as SwiftGen does). If You know how to handle that - please let me know.

File: FantasyUniverse/Scenes/Dashboard/DashboardViewModel.swift (hunk 12 above)
Comment (Peregrinus0304) [reply FantasyUniverse/Scenes/Dashboard/DashboardViewModel.swift]: 😄

File: FantasyUniverse/Scenes/Dashboard/DashboardView.swift (hunk 10 above)
Comment (Peregrinus0304) [reply FantasyUniverse/Scenes/Dashboard/DashboardView.swift]: It would be probably more consistant with the mvvm used on other scenes, but it would have no functionality, only one constant and three variables, so maybe no need for viewModel.

File: FantasyUniverse/Scenes/ResetPassword/ResetPasswordView.swift [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                ScrollView {
+                    VStack {
+                        Spacer()
+                        successAnimation
+                            .frame(width: animationSideSize,
+                                   height: animationSideSize,
+                                   alignment: .center)
+                        infoLabel
+                            .frame(width: animationSideSize,
+                                   height: UIValues.defaultInfoLabelHeight,
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/ResetPassword/ResetPasswordView.swift]: What is the usecase here? You wouldn't constrain label's height in case it would be much longer with future text changes, in other languages or someone's using a bigger font in their system.

File: FantasyUniverse/Scenes/ResetPassword/ResetPasswordView.swift [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+                           contentMode: .scaleToFill)
+            } else {
+                EmptyView()
+            }
+        }
+    }
+    
+    private var infoLabel: some View {
+        Group {
+            if viewModel.isResetSuccessfull {
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/ResetPassword/ResetPasswordView.swift]: Successful - one `l`

File: FantasyUniverse/Scenes/ResetPassword/ResetPasswordViewModel.swift [hunk 22]
```diff
@@ -13,23 +13,22 @@ class ResetPasswordViewModel: ObservableObject {
     private let validator = AuthenticationValidator()
     private let firebaseAuthService = FirebaseAuthService()
     @Published var emailFieldValue = ""
-    @Published var errorMessage = ""
+    @Published var validationErrorMessage = ""
+    @Published var fieldsValid = false
```
Comment (dzoanb) [inline FantasyUniverse/Scenes/ResetPassword/ResetPasswordViewModel.swift]: Again, having a closed set of mutually exclusive states would be much better. So what something like an enum with a case for states like `initial`, `error`, `valid` and appropriate associated values would do is, it would ensure multiple of these values cannot be active at the same time and the combinations of active values are easier to reason about. For example, what does a `fieldsValid` == false, `validationErrowMessage` == "something", `isResetSuccesful` == true mean?

File: FantasyUniverse/Services/Auth/FirebaseAuthService.swift [hunk 23]
```diff
@@ -30,97 +36,159 @@ final class FirebaseAuthService: ObservableObject {
...
                 if let user = user {
-                    print("Got user: \(user)")
                     
-                   let user = User(
-                        uid: user.uid,
-                        displayName: user.displayName,
-                        email: user.email,
-                        refreshToken: user.refreshToken)
-                    self.state = .success(with: user)
+                    self.fetchedUserInfoPublisher(uid: user.uid)
```
Comment (dzoanb) [inline FantasyUniverse/Services/Auth/FirebaseAuthService.swift:44]: To me, looks like a perfect opportunity to do publisher chaining with the `flatMap` :D

File: FantasyUniverse/Services/Auth/FirebaseAuthService.swift [hunk 24]
```diff
@@ -30,97 +36,159 @@ final class FirebaseAuthService: ObservableObject {
...
+                                profileImageURL: user.photoURL,
+                                firstName: userInfo.firstname,
+                                lastName: userInfo.lastname)
+                            self.state = .success(with: user)
+                            self.currentUserUID = user.uid
+                        }
+                        .store(in: &self.subscriptions)
                 } else {
                     self.state = .unauthenticated
+                    self.currentUserUID = nil
```
Comment (dzoanb) [inline FantasyUniverse/Services/Auth/FirebaseAuthService.swift]: Sounding like a broken record probably, but I'd rather have the ID as an associated value to the `authenticated` state and reduce possible combinations of states

File: FantasyUniverse/Services/Auth/FirebaseAuthService.swift [hunk 25]
```diff
@@ -30,97 +36,159 @@ final class FirebaseAuthService: ObservableObject {
...
                             } else {
-                                promise(.failure(NSError(domain: "Invalid User Id", code: 0, userInfo: nil)))
+                                promise(.failure(.noUser))
                             }
                         }
                     }
             }
         }
-        .receive(on: RunLoop.main)
+        .flatMap { [weak self] uid in
```
Comment (dzoanb) [inline FantasyUniverse/Services/Auth/FirebaseAuthService.swift:114]: ✨

File: FantasyUniverse/Services/Auth/FirebaseAuthService.swift [hunk 26]
```diff
@@ -30,97 +36,159 @@ final class FirebaseAuthService: ObservableObject {
...
             }
         }
-        .receive(on: RunLoop.main)
+        .eraseToAnyPublisher()
+    }
+      
+    func uploadProfileImagePublisher(_ image: UIImage) -> AnyPublisher<Void, AuthError> {
+        Deferred {
+            Future { promise in
+                if let uid = self.currentUserUID {
```
Comment (dzoanb) [inline FantasyUniverse/Services/Auth/FirebaseAuthService.swift]: When the fallback is so small I'd prefer guard probably.

File: FantasyUniverse/Services/Auth/FirebaseAuthService.swift [hunk 27]
```diff
@@ -30,97 +36,159 @@ final class FirebaseAuthService: ObservableObject {
...
+                    guard let imageData = image.jpegData(compressionQuality: 0.5) else {
+                        promise(.failure(.cantCompressImage))
+                        return
+                    }
+                    let storageRef = Storage.storage().reference()
+                    let imagesStorageRef = storageRef.child(AuthFirebaseKeys.images.rawValue).child(uid)
+                    imagesStorageRef.putData(imageData, metadata: nil) { (metadata, error) in
```
Comment (dzoanb) [inline FantasyUniverse/Services/Auth/FirebaseAuthService.swift]: Something a little more elegant, but probably overkill would be to have `putData`, `downloadURL` and `commitChanges` separated to their own proxy methods that return a publisher and chain them with Combine instead of a piramid of doom with multiple nested completion blocks

File: FantasyUniverse/Views/App views/UserInfoView.swift [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+import SwiftUI
+
+struct UserInfoView: View {
+    
+    var user: User
+    @Binding var userImage: Image?
+    @State private var showingAvatar = false
+    let logoutAction: () -> Void
+    let editProfileAction: () -> Void
+    let imageTappedAction: () -> Void
```
Comment (dzoanb) [inline FantasyUniverse/Views/App views/UserInfoView.swift]: Looks complex enough for a view model if you ask me 🤔

**Review by dzoanb — APPROVED** [review]
Some comments, nothing ground-breaking.

File: FantasyUniverse/Scenes/ResetPassword/ResetPasswordView.swift (hunk 20 above)
Comment (Peregrinus0304) [reply FantasyUniverse/Scenes/ResetPassword/ResetPasswordView.swift]: Right

File: FantasyUniverse/Services/Auth/FirebaseAuthService.swift (hunk 24 above)
Comment (Peregrinus0304) [reply FantasyUniverse/Services/Auth/FirebaseAuthService.swift]: Right

File: FantasyUniverse/Views/App views/UserInfoView.swift (hunk 28 above)
Comment (Peregrinus0304) [reply FantasyUniverse/Views/App views/UserInfoView.swift]: Hmm... I will probably add a viewModel, but later, as there is be more logic planned.
